packages/components/src/custom-select-control-v2/index.tsx

```diff
@@ -0,0 +1,86 @@
+/**
+ * External dependencies
+ */
+// eslint-disable-next-line no-restricted-imports
+import * as Ariakit from '@ariakit/react';
+/**
+ * WordPress dependencies
+ */
+import { createContext, useContext, useMemo } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import { useCx } from '../utils/hooks/use-cx';
+import * as Styled from './styles';
+import type {
+	CustomSelectProps,
+	CustomSelectItemProps,
+	CustomSelectContext as CustomSelectContextType,
+} from './types';
+
+export const CustomSelectContext =
+	createContext< CustomSelectContextType >( undefined );
+
+export function CustomSelect( props: CustomSelectProps ) {
+	const {
+		children,
+		defaultValue,
+		label,
+		onChange,
+		size = 'default',
+		value,
+		renderSelectedValue,
+	} = props;
+
+	const store = Ariakit.useSelectStore( {
+		setValue: ( nextValue ) => onChange?.( nextValue ),
+		defaultValue,
+		value,
+	} );
+
+	const { value: currentValue } = store.useState();
+
+	const cx = useCx();
+
+	const classes = useMemo(
+		() => cx( Styled.inputSize[ size ] ),
```

Comment:
Nit: since we're already using a styled component for the `CustomSelectButton`, we could avoid also passing a custom classname, and instead move the logic around the `size` prop directly in the styled component.

Reply:
Before I saw this comment, I adjusted the styles [as discussed in an earlier comment](https://github.com/WordPress/gutenberg/pull/55790#discussion_r1397834798) in https://github.com/WordPress/gutenberg/pull/55790/commits/e2ee037d1f8a0c35977224915dca6e0eb97b46a1. 

I tried integrating it into `CustomSelectButton` but ran into many type errors. If you know of an example or docs that might help, I'd love to learn how to accomplish this!

Reply:
Ah just saw [your other comment about this](https://github.com/WordPress/gutenberg/pull/55790#discussion_r1397947198).  I'll read through all of them first instead of one at a time. 😅